ls1intum/Artemis
# `Programming exercises`: Allow instructors to check the consistency of repositories (version control) and build plans (continuous integration)

### Checklist

### Motivation and Context



It might happen that users (or admins) manually edit/remove repositories or build plans, leading to inconsistencies with ArTEMiS. This consistency check feature allows **admins** and **instructors** to check if any of these inconsistencies are present for programming exercises in ArTEMiS. 

This initial feature only includes some checks, but could be extended with further checks in follow-up PRs.

### Description

A new modal component which can be opened with a button in the *exercise detail* and *programming exercises overview* views, has been added. When triggered, a consistency check is performed for an individual programming exercise (or all selected programming exercises) and resulting consistency errors are displayed to the user as a list, if any has been found.

Possible consistency error types are: 

* VCS_PROJECT_MISSING
* TEMPLATE_REPO_MISSING
* SOLUTION_REPO_MISSING
* TEST_REPO_MISSING
* TEMPLATE_BUILD_PLAN_MISSING
* SOLUTION_BUILD_PLAN_MISSING

In case local simulations are tested for consistency, a **IS_LOCAL_SIMULATION** error will be displayed.

### Steps for Testing


1. Log in to Artemis as Tutor / Editor
1.2 Navigate to a programming exercise detail page
1.2.1 The Consistency Check button should not be visible
1.3 Navigate to a course's exercises overview
1.3.1 Select one or more programming exercises with the left checkboxes (The Consistency Check button should not be visible)
2.  Log in to Artemis as Instructor / Admin
2.1 Navigate to a course containing programming exercises
2.2 Create a new programming exercise
2.3 Remove from Bitbucket/Gitlab one of the repositories (or buid plan from Bamboo/Jenkins) of the generated exercise
2.4 Navigate to the course's exercises overview, select one or more programming exercises (including the one without build plan/repository and click the **Check Consistency button**
2.4.1 An error informing about the missing repository/build plan should appear

### Test Coverage



ConsistencyCheckService: 100%
ProgrammingExerciseService: 92%
ConsistencyCheckResource: 100%

course-detail.component.ts: 91%
consistency-check-result.model.ts: 100%
programming-exercise-detail.component.ts: 48%
consistency-check.component.ts: 83%
consistency-check.service.ts: 80%

### Screenshots


![Screenshot from 2021-08-03 15-51-16](https://user-images.githubusercontent.com/45761921/128027272-99081bec-d3fa-42d8-a9ad-184b15fcf0d2.png)

![image](https://user-images.githubusercontent.com/45761921/139453393-ff657fdb-a8ef-46b1-ac68-4a10f254e50d.png)
